code/z_adventurezones/space_casino.dm

```diff
@@ -0,0 +1,371 @@
+					explode()
+					return
+				if (get_turf(src) == get_turf(home)) return
+				src.navigate_to(get_turf(home), FLOORBOT_MOVE_SPEED, max_dist = 10)
+				if (!src.path || !length(src.path))
+					KillPathAndGiveUp(1)
+					return
+				return
+		// Something to do. Let's decide what to do first.
+		if (!target)
+			target = targets[1]
+		// Now we need to find out if we're going to move towards a prop dispenser or the table.
+		if (!moveTowards)
+			if (!hasDrink)
+				// if there's a barbuddy dispenser nearby, let's do the cute little animation thing. if not, use magic to summon a drink
+				for (var/obj/decal/fakeobjects/barbuddy_dispenser/D in view(5, src))
+					moveTowards = D
+				if (!moveTowards)
+					hasDrink = 1
+			else
+				moveTowards = targets[1]
+
+		if (IN_RANGE(get_turf(src), get_turf(src.moveTowards), 1))
+			bartend()
+			return
+
+		if (isnull(get_turf(moveTowards)))
+			if (moveTowards in src.homeTables)
+				src.homeTables -= moveTowards
+			KillPathAndGiveUp(1)
+			return
+
+		if (!src.path || !length(src.path))
+			src.navigate_to(get_turf(moveTowards), FLOORBOT_MOVE_SPEED, max_dist = 10)
+			if (!src.path || !length(src.path))
+				KillPathAndGiveUp(1)
+				return
+
+	KillPathAndGiveUp(var/give_up)
+		. = ..()
+		if(give_up)
+			src.targets -= src.target
+			src.target = null
+			moveTowards = null
+			// Let's check if we've been stolen.
+			if (!(src.home in view(5, src)))
+				explode()
+
+	proc/bartend()
+		if (istype(moveTowards, /obj/decal/fakeobjects/barbuddy_dispenser)) // If it's the dispenser, do a little animation.
+			playsound(moveTowards.loc, 'sound/misc/pourdrink2.ogg', 50, 1, 0.3)
+			moveTowards.icon_state = "alc_dispenser[rand(1,5)]"
+			hasDrink = 1
+			moveTowards = null
+		else if (istype(moveTowards, /obj/table/reinforced/bar)) // If it's a table, so let's generate a drink.
+			var/pickedVessel = pick(possible_vessels)
+			var/obj/item/reagent_containers/food/drinks/drinkingglass/W = new pickedVessel(moveTowards.loc)
+			if (src.emagged)
+				W.reagents.add_reagent(pick(possible_poisons), W.initial_volume)
+			else
+				W.reagents.add_reagent(pick(possible_drinks), W.initial_volume)
+			W.pixel_x = rand(-8, 8)
+			W.pixel_y = rand(0, 16)
+			if (prob(25)) // Chance of stuff!
+				var/pickedStuff = pick(possible_stuffs)
+				var/obj/item/cocktail_stuff/U = new pickedStuff(null)
+				W.in_glass = U
+				W.update_icon()
+			if (prob(25)) // Chance of wedge!
+				var/pickedWedge = pick(possible_wedges)
+				var/obj/item/reagent_containers/food/snacks/plant/P = new pickedWedge(null)
+				W.wedge = P
+				W.update_icon()
+			hasDrink = 0
+			KillPathAndGiveUp(1)
+
+	emag_act(var/mob/user, var/obj/item/card/emag/E)
+		if (user)
+			user.show_text("You show [src] your [E]. They smile so hard that they begin sparking!", "red")
+		emagged = 1
+
+	demag(var/mob/user)
+		emagged = 0
+
+	explode()
+		src.visible_message("<span class='alert'><B>[src] gets confused and explodes!</B></span>", 1)
```

Comment:
Also note that explode() is called on normal death, so unless you want them to be confused whenever someone clubs it to death, might wanna put an exception for when it gets homesick.